### PR TITLE
Removed queries against information_schema

### PR DESCRIPTION
**Pull request by allouis**
no-issue

We were originally checking the state of the database, e.g. if a foreign
key constraint existed, so that we could conditionally act upon it. This
was to ensure that our migrations are idempotent.

Some database configurations, for example if you have many databases on
a single MySQL instance, would cause these information_schema queries to
take an exceptionally long time.

In order to speed up migrations, we instead attempt the action we want
to apply to the database, and then catch relevant errors to ensure the
migration is idempotent.

The errors we catch are all MySQL specific, excluding the SQLITE_ERROR
which seems to only be thrown when working with unique constraints

### CODE REVIEW

**Review by naz — CHANGES_REQUESTED** [review]
Tested this for MySQL and SQLite after running `DELETE FROM migrations WHERE version = '4.0';`

<details>
<summary>Log produced by MySQL</summary>

[2021-03-09 01:25:39] INFO Adding version to posts.mobiledoc objects and cleaning up deprecated card names
[2021-03-09 01:25:39] WARN Adding members.status column - skipping as table is correct
[2021-03-09 01:25:39] INFO Updating members.status based on members_stripe_customers_subscriptions.status
[2021-03-09 01:25:39] WARN Failed to drop table: permissions_apps - table does not exist
[2021-03-09 01:25:39] WARN Failed to drop table: app_fields - table does not exist
[2021-03-09 01:25:39] WARN Failed to drop table: app_settings - table does not exist
[2021-03-09 01:25:39] WARN Failed to drop table: apps - table does not exist
[2021-03-09 01:25:39] WARN Skipping adding table: members_subscribe_events - table already exists
[2021-03-09 01:25:39] INFO Populating members_subscribe_events from members
[2021-03-09 01:25:39] INFO Dropping unique constraint for: slug in table: posts
[2021-03-09 01:26:08] WARN Constraint for: slug does not exist for table: posts
[2021-03-09 01:26:08] INFO Adding unique constraint for: slug,type in table posts
[2021-03-09 01:26:16] WARN Constraint for: slug,type already exists for table: posts
[2021-03-09 01:26:16] WARN Skipping adding table: members_login_events - table already exists
[2021-03-09 01:26:16] WARN Skipping adding table: members_email_change_events - table already exists
[2021-03-09 01:26:16] WARN Skipping adding table: members_status_events - table already exists
[2021-03-09 01:26:16] WARN Skipping adding table: members_paid_subscription_events - table already exists
[2021-03-09 01:26:16] INFO Removing installed_apps,active_apps from "settings" table.
[2021-03-09 01:26:16] WARN Skipping adding table: members_payment_events - table already exists
[2021-03-09 01:26:16] INFO Skipping removal of orphaned stripe records for MySQL
[2021-03-09 01:26:16] WARN Adding posts_meta.frontmatter column - skipping as table is correct
[2021-03-09 01:26:16] INFO Populating slack_url and slack_username setting values
[2021-03-09 01:26:16] WARN slack_url and slack_username settings already exist
[2021-03-09 01:26:16] INFO Populating members_status_events from members table
[2021-03-09 01:26:16] INFO Transforming all internal urls to transform-ready
[2021-03-09 01:26:16] INFO Deleting labs from settings table
[2021-03-09 01:26:16] INFO Updating unsplash setting value
[2021-03-09 01:26:16] WARN Skipping update of unsplash value. Current value is already boolean: true
[2021-03-09 01:26:16] INFO Sanitizing provider_id values in email_batches
[2021-03-09 01:26:16] INFO Resolving the orphaned webhooks
[2021-03-09 01:26:16] INFO No orphaned webhooks found, skipping
[2021-03-09 01:26:16] INFO Starting re-generation of posts html.
[2021-03-09 01:26:16] INFO Finished re-generation of posts html.
[2021-03-09 01:26:16] WARN Permission for read:email already added
[2021-03-09 01:26:16] WARN Adding permission(Read emails) to role(Administrator) - already exists
[2021-03-09 01:26:16] WARN Adding permission(Read emails) to role(Admin Integration) - already exists
[2021-03-09 01:26:16] WARN Adding permission(Read emails) to role(Editor) - already exists
[2021-03-09 01:26:16] WARN Adding permission(Read emails) to role(Author) - already exists
[2021-03-09 01:26:16] WARN Adding permission(Read emails) to role(Contributor) - already exists
[2021-03-09 01:26:16] WARN Permission for browse:email already added
[2021-03-09 01:26:16] WARN Adding permission(Browse emails) to role(Administrator) - already exists
[2021-03-09 01:26:16] WARN Adding permission(Browse emails) to role(Admin Integration) - already exists
[2021-03-09 01:26:16] WARN Adding permission(Browse emails) to role(Editor) - already exists
[2021-03-09 01:26:16] WARN Permission for retry:email already added
[2021-03-09 01:26:16] WARN Adding permission(Retry emails) to role(Administrator) - already exists
[2021-03-09 01:26:16] WARN Adding permission(Retry emails) to role(Admin Integration) - already exists
[2021-03-09 01:26:16] WARN Adding permission(Retry emails) to role(Editor) - already exists
[2021-03-09 01:26:17] INFO Populating members_paid_subscription_events from members_stripe_customers_subscriptions
[2021-03-09 01:26:17] WARN Skipping adding "on delete cascade" - database is not SQLite3
[2021-03-09 01:26:17] INFO Adding a primary key for the brute table
[2021-03-09 01:26:17] INFO Adding primary key constraint for: key in table brute
[2021-03-09 01:26:26] WARN Primary key constraint for: key already exists for table: brute
[2021-03-09 01:26:26] INFO Adding the webhooks to integrations foreign key
[2021-03-09 01:26:26] INFO Adding foreign key from webhooks.integration_id to integrations.id
[2021-03-09 01:26:32] WARN Skipped adding foreign key from webhooks.integration_id to integrations.id - foreign key already exists

</details>

<details>
<summary>Log produced by SQLite</summary>

[2021-03-09 01:34:27] INFO Adding version to posts.mobiledoc objects and cleaning up deprecated card names
[2021-03-09 01:34:27] WARN Adding members.status column - skipping as table is correct
[2021-03-09 01:34:27] INFO Updating members.status based on members_stripe_customers_subscriptions.status
[2021-03-09 01:34:27] WARN Failed to drop table: permissions_apps - table does not exist
[2021-03-09 01:34:27] WARN Failed to drop table: app_fields - table does not exist
[2021-03-09 01:34:27] WARN Failed to drop table: app_settings - table does not exist
[2021-03-09 01:34:27] WARN Failed to drop table: apps - table does not exist
[2021-03-09 01:34:27] WARN Skipping adding table: members_subscribe_events - table already exists
[2021-03-09 01:34:27] INFO Populating members_subscribe_events from members
[2021-03-09 01:34:27] INFO Dropping unique constraint for: slug in table: posts
[2021-03-09 01:34:27] WARN Constraint for: slug does not exist for table: posts
[2021-03-09 01:34:27] INFO Adding unique constraint for: slug,type in table posts
[2021-03-09 01:34:30] WARN Constraint for: slug,type already exists for table: posts
[2021-03-09 01:34:30] WARN Skipping adding table: members_login_events - table already exists
[2021-03-09 01:34:30] WARN Skipping adding table: members_email_change_events - table already exists
[2021-03-09 01:34:30] WARN Skipping adding table: members_status_events - table already exists
[2021-03-09 01:34:30] WARN Skipping adding table: members_paid_subscription_events - table already exists
[2021-03-09 01:34:30] INFO Removing installed_apps,active_apps from "settings" table.
[2021-03-09 01:34:30] WARN Skipping adding table: members_payment_events - table already exists
[2021-03-09 01:34:30] INFO Removing orphaned rows from members_stripe_customers
[2021-03-09 01:34:30] INFO Removing orphaned rows from members_stripe_customers_subscriptions
[2021-03-09 01:34:30] WARN Adding posts_meta.frontmatter column - skipping as table is correct
[2021-03-09 01:34:30] INFO Populating slack_url and slack_username setting values
[2021-03-09 01:34:30] WARN slack_url and slack_username settings already exist
[2021-03-09 01:34:30] INFO Populating members_status_events from members table
[2021-03-09 01:34:30] INFO Transforming all internal urls to transform-ready
[2021-03-09 01:34:30] INFO Deleting labs from settings table
[2021-03-09 01:34:31] INFO Updating unsplash setting value
[2021-03-09 01:34:31] WARN Skipping update of unsplash value. Current value is already boolean: true
[2021-03-09 01:34:31] INFO Sanitizing provider_id values in email_batches
[2021-03-09 01:34:31] INFO Resolving the orphaned webhooks
[2021-03-09 01:34:31] INFO No orphaned webhooks found, skipping
[2021-03-09 01:34:31] INFO Starting re-generation of posts html.
[2021-03-09 01:34:31] INFO Finished re-generation of posts html.
[2021-03-09 01:34:31] WARN Permission for read:email already added
[2021-03-09 01:34:31] WARN Adding permission(Read emails) to role(Administrator) - already exists
[2021-03-09 01:34:31] WARN Adding permission(Read emails) to role(Admin Integration) - already exists
[2021-03-09 01:34:31] WARN Adding permission(Read emails) to role(Editor) - already exists
[2021-03-09 01:34:31] WARN Adding permission(Read emails) to role(Author) - already exists
[2021-03-09 01:34:31] WARN Adding permission(Read emails) to role(Contributor) - already exists
[2021-03-09 01:34:31] WARN Permission for browse:email already added
[2021-03-09 01:34:31] WARN Adding permission(Browse emails) to role(Administrator) - already exists
[2021-03-09 01:34:31] WARN Adding permission(Browse emails) to role(Admin Integration) - already exists
[2021-03-09 01:34:31] WARN Adding permission(Browse emails) to role(Editor) - already exists
[2021-03-09 01:34:31] WARN Permission for retry:email already added
[2021-03-09 01:34:31] WARN Adding permission(Retry emails) to role(Administrator) - already exists
[2021-03-09 01:34:31] WARN Adding permission(Retry emails) to role(Admin Integration) - already exists
[2021-03-09 01:34:31] WARN Adding permission(Retry emails) to role(Editor) - already exists
[2021-03-09 01:34:31] INFO Populating members_paid_subscription_events from members_stripe_customers_subscriptions
[2021-03-09 01:34:31] INFO Adding on delete cascade for members_labels, members_stripe_customers and members_stripe_customers_subscriptions
[2021-03-09 01:34:31] INFO Dropping foreign key from members_labels.member_id to members.id
[2021-03-09 01:34:31] INFO Adding foreign key from members_labels.member_id to members.id
[2021-03-09 01:34:31] INFO Dropping foreign key from members_labels.label_id to labels.id
[2021-03-09 01:34:31] INFO Adding foreign key from members_labels.label_id to labels.id
[2021-03-09 01:34:31] INFO Dropping foreign key from members_stripe_customers.member_id to members.id
[2021-03-09 01:34:31] INFO Adding foreign key from members_stripe_customers.member_id to members.id
[2021-03-09 01:34:31] INFO Dropping foreign key from members_stripe_customers_subscriptions.customer_id to members_stripe_customers.id
[2021-03-09 01:34:31] INFO Adding foreign key from members_stripe_customers_subscriptions.customer_id to members_stripe_customers.id
[2021-03-09 01:34:31] INFO Adding a primary key for the brute table
[2021-03-09 01:34:31] INFO Adding primary key constraint for: key in table brute
[2021-03-09 01:34:31] INFO Adding the webhooks to integrations foreign key
[2021-03-09 01:34:31] INFO Adding foreign key from webhooks.integration_id to integrations.id

</details>


Notice SQLite version did not throw errors and continued creating duplicate FKs and PKs. After running migrations multiple times on SQLite have this state of `webhooks` table:
```
sqlite> PRAGMA foreign_key_list('webhooks');
0|0|integrations|integration_id|id|NO ACTION|CASCADE|NONE
1|0|integrations|integration_id|id|NO ACTION|CASCADE|NONE
2|0|integrations|integration_id|id|NO ACTION|CASCADE|NONE
3|0|integrations|integration_id|id|NO ACTION|CASCADE|NONE
```

We probably need to leave some of the handling at least for SQLite? :thinking:

File: core/server/data/schema/commands.js [hunk 1]
```diff
@@ -178,12 +213,20 @@ async function addForeign({fromTable, fromColumn, toTable, toColumn, cascadeDele
...
+        }
+    }
     try {
         logging.info(`Dropping foreign key from ${fromTable}.${fromColumn} to ${toTable}.${toColumn}`);
 
         //disable and re-enable foreign key checks on sqlite because of https://github.com/knex/knex/issues/4155
         let foreignKeysEnabled;
-        if (db.knex.client.config.client === 'sqlite3') {
+        if (isSQLite) {
             foreignKeysEnabled = await db.knex.raw('PRAGMA foreign_keys;');
```
Comment (allouis) [inline core/server/data/schema/commands.js:230]: Hey @tpatel - are we purposefully not using the passed in transaction/connection here, and opting for using `db.knex` directly?

Comment (tpatel) [reply core/server/data/schema/commands.js:224]: Not purposefully, so this change is fine!

Comment (allouis) [reply core/server/data/schema/commands.js:224]: I think that can be done as part of cleanup tho IMO

Comment (tpatel) [reply core/server/data/schema/commands.js:224]: I meant I think the transaction/knex instance should give the same client as db.knex.client.config.client, right?

File: core/server/data/schema/commands.js [hunk 2]
```diff
@@ -258,64 +231,67 @@ async function dropForeign({fromTable, fromColumn, toTable, toColumn, transactio
...
+ * @param {import('knex')} transaction - connnection object containing knex reference
  */
 async function addPrimaryKey(tableName, columns, transaction) {
-    const hasUniqueConstraint = await hasPrimaryKey(tableName, transaction);
-
-    if (!hasUniqueConstraint) {
+    const isSQLite = db.knex.client.config.client === 'sqlite3';
+    if (isSQLite) {
+        const primaryKeyExists = await hasPrimaryKeySQLite(tableName, transaction);
+        if (!primaryKeyExists) {
```
Comment (daniellockyer) [inline core/server/data/schema/commands.js:279]: @allouis Should this be:

```suggestion
        if (primaryKeyExists) {
```

**Review by daniellockyer — APPROVED** [review]

**Comment by tpatel** [comment]
good for me 👍

**Pull request merged**